Add a target for Vitis IP blocks and fix some performance issues

Ralender opened the pull request:
to target a vitis ip block the command looks like:
`clang++ --target=vitis_ip-xilinx vitis_ip.cpp -std=c++20 -o a.zip --vitis-ip-part=xc7vx330t-ffg1157-1`
or to get fast performance reports on SYCL device code. the usual command + `--vitis-ip-part=xc7vx330t-ffg1157-1` like:
`clang++ -fsycl -fsycl-targets=fpga64_hls_hw_emu test.cpp -std=c++20 --vitis-ip-part=xc7vx330t-ffg1157-1`

This path also contain improvement to the optimization pipeline used on our side for device code:
 - improve promotion of (virtual)memory to (virtual)registers which should:
      - reduce the frequency of pointer on pointer which are problematic for HLS.
      - prevent weird bit width integers like i23 or i6 from being rounded up to i24 or i8.
 - improve the lowering of memory intrinsic which used to confuse HLS into making horrible HLD.

Commit:
[SYCL][VXX] First attempt at trying to find a middle ground between running -O3 and running nothing in sycl_vxx

Commit:
[SYCL][VXX][SROA] Improve filtering of what alloc should be handled by our SROA or HLS's SROA

Commit:
[SYCL][VXX] Handle XILINX_CLANG_39_BUILD_PATH correctly in sycl_vxx.py

Commit:
[SYCL][VXX] Fixed intrisic lowering

Commit:
[VXX][VitisIP] First attempt of adding vitis_hls ip flow

Commit:
[VXX] Add vitis_ip target to build block design ip

Commit:
[VXX][VitisIP] Fixed the flow

Commit:
[VXX][VitisIP] Add tests for vitis ip target

allow targetting vitis ip when compilling sycl code.
use C name instead of C++ mangled names for the top-level function in vitis kernels

Commit:
[VXX][SYCL][VitisIP] Fix some driver issues and cleanup code

Commit:
[VXX] fix issue

keryell commented:
Perhaps this needs a rebase after the merge of https://github.com/triSYCL/sycl/pull/156

Commit:
[VXX] Fix sycl_vxx.py

Commit:
[VXX] Fix issue preventing the driver to find clang tools.

Commit:
[VXX][SYCL] Refactor memory intrinsic lowering.

Commit:
[VXX][SYCL] Add support for --vitis-ip-part during the SYCL compilation flow.

This has the effect of compiling the SYCL device code with the ip export flow which is much faster
than the V++ flow because it compiles only the kernel and not the kernel + the shell.

/!\ BUT the resulting executable WILL NOT WORK because it has not shell /!\

so this flow only exists as a very fast way to get performance reports for SYCL kernel code.

Ralender commented:
updated description

Commit:
[VXX] Refactor sycl_vxx to prevent processing HLS's LLVM IR

Commit:
[VXX][SYCL] Fix some failing test

Commit:
[VXX] Fix DDR bank IR reprentation

keryell commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4769,6 +4769,15 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+    /// -O3 to generate all the necessary information for proper optimization.
+    CmdArgs.push_back("-O3");
+    /// -disable-llvm-passes because we want sycl_vxx and v++ to have full
+    /// control over the IR. so we disable any optimization that could run
```
```suggestion
    /// Use -disable-llvm-passes because we want sycl_vxx and v++ to have full
    /// control over the IR, so we disable any optimization that could run
```

keryell commented on clang/lib/Driver/ToolChains/VXX.cpp:
```diff
@@ -138,16 +100,19 @@ void AddForwardedOptions(const llvm::opt::ArgList &Args,
 void SYCL::LinkerVXX::constructSYCLVXXCommand(
     Compilation &C, const JobAction &JA, const InputInfo &Output,
     const InputInfoList &Inputs, const llvm::opt::ArgList &Args) const {
+
```
?

keryell commented on clang/lib/Driver/ToolChains/VXX.cpp:
```diff
@@ -138,16 +100,19 @@ void AddForwardedOptions(const llvm::opt::ArgList &Args,
-  assert(!TC.VXXInstallation.getBinPath().empty());
-  CmdArgs.push_back("--vitis_bin_dir");
-  CmdArgs.push_back(Args.MakeArgString(TC.VXXInstallation.getBinPath()));
+  bool isVitisIP = TC.isVitisIP() || Args.hasArg(options::OPT_vitis_ip_part_EQ);
```
You should not remove comments but add some or fix the previous ones :-)

keryell commented on llvm/lib/SYCL/LowerSYCLMetaData.cpp:
```diff
@@ -310,6 +314,15 @@ struct LSMDState {
     }
   }
 
+  void markKernel(llvm::ConstantStruct *CS) {
```
Add top comment

keryell commented on llvm/lib/SYCL/LowerSYCLMetaData.cpp:
```diff
@@ -164,13 +164,17 @@ struct LSMDState {
   }
 
   void lowerPipelineKernelDecoration(llvm::Function *F, llvm::Value *Payload) {
```
While you are here, add a top level comment

keryell commented on llvm/lib/SYCL/PrepareSYCLOpt.cpp:
```diff
@@ -244,11 +247,34 @@ struct PrepareSYCLOpt : public ModulePass {
     }
   }
 
+  void lowerMemIntrinsic(Module &M) {
```
Top-level description

keryell commented on llvm/lib/SYCL/VXXIRDowngrader.cpp:
```diff
@@ -232,38 +235,31 @@ struct VXXIRDowngrader : public ModulePass {
-  }
-
-  void lowerIntrinsic(Module &M) {
+  void lowerAbsIntrinsic(Module &M) {
```
Could this be generalized with the memory intrinsics lowering?
+ top-level comments.

Ralender replied:
the memory intrinsic are lowered elsewhere because we need optimizations after they are lowered.

keryell commented on llvm/lib/SYCL/VXXIRDowngrader.cpp:
```diff
@@ -232,38 +235,31 @@ struct VXXIRDowngrader : public ModulePass {
+          ToProcess.push_back(&I);
+    for (auto *I : ToProcess) {
+      if (auto *CI = dyn_cast<CallBase>(I)) {
+        if (CI->getIntrinsicID() == Intrinsic::abs) {
```
Simplify (a little) with early exit with `!= Intrinsic::abs` so you can erase in the else block and remove some `{}`

keryell commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -116,8 +116,11 @@ STATISTIC(NumVectorized, "Number of vectorized aggregates");
 
-static cl::opt<bool> SROAAvoidAggregates("sroa-avoid-aggregates", cl::init(false),
-                                         cl::Hidden);
+/// Prevent SROA from desegregating structs containing variable width integers
```
```suggestion
/// Prevent SROA from desegregating structs containing arbitrary-width integers
```
There is no variation over time in the width itself.

keryell commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -116,8 +116,11 @@ STATISTIC(NumVectorized, "Number of vectorized aggregates");
-static cl::opt<bool> SROAAvoidAggregates("sroa-avoid-aggregates", cl::init(false),
-                                         cl::Hidden);
+/// Prevent SROA from desegregating structs containing variable width integers
+/// with non-common bits sizes.
```
```suggestion
/// with non-common bit sizes.
```

keryell commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -10,10 +10,88 @@
+
+cl::opt<bool>
+    LowerToNonI8Type("lower-mem-intr-to-llvm-type", cl::Hidden,
+                       cl::desc("try to lower memory intrinsics to their "
```
Curious indent

Ralender replied:
Thats what clang-format chose

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -83,127 +184,124 @@ def _run_in_isolated_proctree(cmd, *args, **kwargs):
 
 
+# This is currently unused because a change in from 2021.2 to 2022.1 was reverted
+#  but it is likely to become usefull in the future so i leave it as is
```
```suggestion
# This is currently unused because a change between version 2021.2 and 2022.1 was later reverted.
#  But it is likely to become useful again in the future, so keep it as is
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -238,81 +341,135 @@ def _downgrade(self):
+                print("Please edit", self.downgraded_ir)
+                input("Press enter to resume the compilation")
+            self.vpp_llvm_input = (
+            )
```
```suggestion
            self.vpp_llvm_input = ()
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -238,81 +341,135 @@ def _downgrade(self):
+        with tmp_manager as self.tmpdir:
+            tmpdir = self.tmpdir
+            if not autodelete:
+                print(f"Temporary clutter in {tmpdir} will not be deleted")
```
Isn't it a message which should be handled by the TmpDirManager directly?

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -238,81 +341,135 @@ def _downgrade(self):
+            command = [
+                vxx, "--target", self.vitis_mode,
+                "--advanced.param", "compiler.hlsDataflowStrictMode=off",
+                # Do the optimizations that were not performed by sycl compiler
```
```suggestion
                # Do the optimizations that were not performed by the SYCL compiler
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -351,78 +508,126 @@ def _link_kernels(self):
+
+
+def parse_args(args=sys.argv[1:]):
+    description="Utility to drive various compilation flow vor vivado related tools"
```
```suggestion
    description="Utility to drive various compilation flow for Vivado-related tools"
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -351,78 +508,126 @@ def _link_kernels(self):
+    if command == "vxxcompile":
+        parser.add_argument(
+            "--hls",
+            help="Activate the hls flow instead of the default spir one",
```
```suggestion
            help="Activate the HLS flow instead of the default SPIR one",
```

keryell reviewed (changes requested):
Very interesting!
Use a spell-checker also in commit messages: `[[VXX] Fix DDR bank IR reprentation]` (https://github.com/triSYCL/sycl/pull/162/commits/33c3941948599d489e3f03b8d27add63dbd8af8d)

Commit:
Address review comment

Commit:
[VXX] Fix partion_array.

I cant know if partion array does what it is expected to do
but i changed the le ressources usage as we expect so i assume it works.

Commit:
[VitisIP] Add documentation and simplify usage

lforg37 committed:
Fix some typos in new compiler flow documentation

Ralender committed:
Update GettingStartedIPblocksDeisgn.md

lforg37 committed:
[VXX] Edit IP Block flow documentation 

Change the git remote given in the doc from ssh to https to avoid permission issue

Commit:
[SYCL][VXX] Add N-dimension support to partition_array and improve the API

Also add documentation for extentions

Ralender committed:
Update GettingStartedIPblocksDeisgn.md

Add some more details based on feedbacks and testing.

Commit:
[VXX][lit] Fix unreliable tests

keryell commented on llvm/lib/SYCL/VXXIRDowngrader.cpp:
```diff
@@ -17,24 +17,24 @@
 
 #include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/SetVector.h"
```
Sort

keryell commented on llvm/lib/SYCL/VXXIRDowngrader.cpp:
```diff
@@ -17,24 +17,24 @@
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
+#include "llvm/SYCL/VXXIRDowngrader.h"
```
Sort after ` #include "llvm/Pass.h"`?

keryell commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -10,10 +10,88 @@
+    if (BitWidth <= 64)
+      return llvm::ConstantInt::get(Ty, IntValue);
+    return llvm::ConstantInt::get(
+        Ty, llvm::APInt::getSplat(BitWidth, llvm::APInt(64, IntValue)));
```
I am lost in this area. What about adding some comments inside this huge function about what is happening?

Ralender replied:
The details of the function aren't important. but I added a top-level comment explain what the function does.

keryell commented on sycl/doc/GettingStartedIPblocksDeisgn.md:
```diff
@@ -0,0 +1,61 @@
+Getting started with Vitis ip block design with an AMD/Xilinx FPGA
+===================================================================================
```
```suggestion
Getting started with Vitis IP block design with an AMD/Xilinx FPGA
===================================================================================
```

keryell commented on sycl/doc/GettingStartedIPblocksDeisgn.md:
```diff
@@ -0,0 +1,61 @@
+
+### Small examples
+
+create a file named vitis_ip.cpp containing
```
```suggestion
create a file named `vitis_ip.cpp` containing
```

keryell commented on sycl/doc/GettingStartedIPblocksDeisgn.md:
````diff
@@ -0,0 +1,61 @@
+./build/bin/clang++ --target=vitis_ip-xilinx vitis_ip.cpp --vitis-ip-part=xc7vx330t-ffg1157-1 -o adder.zip
+```
+
+ ``--target=vitis_ip-xilinx`` specifies that we are targeting vitis ip blocks
````
```suggestion
 `--target=vitis_ip-xilinx` specifies that we are targeting Vitis IP blocks
```
Do not get confused between MarkDown and ReST format

keryell commented on sycl/doc/GettingStartedIPblocksDeisgn.md:
````diff
@@ -0,0 +1,61 @@
+```
+
+ ``--target=vitis_ip-xilinx`` specifies that we are targeting vitis ip blocks
+ ``--vitis-ip-part=xc7vx330t-ffg1157-1`` specifies which Xilinx device we are targeting ``xc7vx330t-ffg1157-1`` needs to be replaced by the part id you are working with.
````
```suggestion
 `--vitis-ip-part=xc7vx330t-ffg1157-1` specifies which Xilinx device we are targeting: `xc7vx330t-ffg1157-1` needs to be replaced by the part id you are working with.
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+a few SYCL extensions are provided as a replacement of Vitis HLS `HLS` pragmas.
+
+These extensions requires the inclusion of the `sycl/ext/xilinx/fpga.hpp`
+header, and requires compiling with C++20 support.
```
```suggestion
header, and requires compiling with support of C++20 or more.
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+
+Two types are also defined to describe more specifically loop pipeline behavior:
+
++ `NoRewindPipeline` loop pipelines are flushed between two execution of the full loop, while
```
I think there is a coding style problem here.
Should not it be `no_rewind_pipeline` and so on?

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+chh.single_task([=]{
+  // Pipeline the following loop
+  for (size_t i = 0 ; i < 42 ; ++i>) {
+    pipeline<>([&]{
```
It would be nice to be able to just write
```suggestion
    pipeline([&]{
```
Using CTAD by having `pipeline` a class instead of a function?

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+Kernel pipelining:
+
+```cpp
+cgh.single_task(pipeline_kernel<ConstrainedII<4>>(
````
```suggestion
cgh.single_task(pipeline_kernel<constrained_ii<4>>(
```
or with a more CTAD spirit
```suggestion
cgh.single_task(pipeline_kernel { constrained_ii<4>, [=] { ...
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+### Dataflow decorators
+
+These decorators are the equivalent of the
+[HLS dataflow pragma](https://www.xilinx.com/html_docs/xilinx2021_1/vitis_doc/hls_pragmas.html#sxx1504034358866).
```
```suggestion
[HLS dataflow pragma](https://docs.xilinx.com/r/en-US/ug1399-vitis-hls/pragma-HLS-dataflow).
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+
+```cpp
+sycl::queue queue;
+sycl::buffer<int, 42> a_buff, b_buff, c_buff;
````
Do you really want a buffer with 42 dimensions? I have a prototype for this https://github.com/triSYCL/triSYCL/pull/313 :-)

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+  sycl::ext::oneapi::accessor_property_list property_bank_2{sycl::ext::xilinx::ddr_bank<2>};
+  sycl::ext::oneapi::accessor_property_list property_bank_3{sycl::ext::xilinx::ddr_bank<3>};
+
+  const sycl::accessor a_acc{a_buff, cgh, sycl::read_only, property_bank_1};
```
This is super verbose.
Do you think it could be possible to write
```suggestion
  const sycl::accessor a_acc{a_buff, cgh, sycl::read_only, {sycl::ext::xilinx::ddr_bank<1>}};
```
with some tricks?
Or even having an overload taking just one property?
```suggestion
  const sycl::accessor a_acc{a_buff, cgh, sycl::read_only, sycl::ext::xilinx::ddr_bank<1>};
```

Ralender replied:
it is possible by adding a new overload to accessors (which already has too many).
or by providing template variables that are already of type accessor_property_list but this would prevent having multiple properties. per accessors.

but otherwise the constructor cannot deduce that the initilizer_list is a accessor_property_list.

alternatively we can do ```const sycl::accessor a_acc{a_buff, cgh, sycl::read_only, sycl::ext::oneapi::accessor_property_list{sycl::ext::xilinx::ddr_bank<1>}};```

but it isn't much better.

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+```
+
+In this example, `a_buff` copy on device will be stored on DDR 
+bank 1, `b_buff` copy on DDR bank 2 and `c_buff` copy on 
````
It is not really a copy, I do not know exactly what we can use
```suggestion
In this example, `a_buff` materialization on device will be stored on DDR
bank 1, `b_buff` materialization on DDR bank 2 and `c_buff` materialization on 
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+### Array partitioning
+
+Partitioning an array allows having faster accesses to that array while using more hardware.
+all code will be shown with
+```cpp
+using namespace sycl::ext::xilinx
+```
````
Is this some spurious old code?

Ralender replied:
well it depends how you see it. most of the code in this doc will not compile because the namespace are not specified.
and the namespaces are not specified because the code becomes way too verbose

but at least stating that the code needs these namespace is the minimum to not confuse users

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+
+`partition_ndarray<int, sycl::dim<2, 3>, partition::complete<>>` is equivalent to the C-style array `int array[2][3]` but we specified a complete partitioning like the [HLS array_partition pragma](https://www.xilinx.com/htmldocs/xilinx2021_1/vitis_doc/hls_pragmas.html#gle1504034361378).
+
++ `partition::complete<>` each element will be in an individual register
```
```suggestion
+ `partition::complete` each element will be in an individual register
```

Ralender replied:
the `<>` cannot be removed

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+using namespace sycl::ext::xilinx
+```
+
+`partition_ndarray<int, sycl::dim<2, 3>, partition::complete<>>` is equivalent to the C-style array `int array[2][3]` but we specified a complete partitioning like the [HLS array_partition pragma](https://www.xilinx.com/htmldocs/xilinx2021_1/vitis_doc/hls_pragmas.html#gle1504034361378).
````
The `nd` does not add that much, even if it is different from `std::mdarray` proposal.
Could it be possible to simplify as
```suggestion
`partition_array<int, sycl::dim<2, 3>, partition::complete>` is equivalent to the C-style array `int array[2][3]` but we specified a complete partitioning like the [HLS array_partition pragma](https://docs.xilinx.com/r/en-US/ug1399-vitis-hls/pragma-HLS-array_partition).
```
or even using concepts in a variadic context
```suggestion
`partition_array<int, 2, 3, partition::complete>` is equivalent to the C-style array `int array[2][3]` but we specified a complete partitioning like the [HLS array_partition pragma](https://docs.xilinx.com/r/en-US/ug1399-vitis-hls/pragma-HLS-array_partition).
```

Ralender replied:
if we remove the `nd` we cannot have an overload for 1d partition_array (which probably the most frequent) that avoid needing to write the sycl::dim

if this isnt an issue we can just use partition_array everywhere

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+Example:
+```cpp
+sycl::queue queue;
+sycl::buffer<int, 42> a_buff, b_buff;
````
```suggestion
auto constexpr N = 42;
sycl::buffer<int> a_buff{N}, b_buffN};
```
I guess all these code samples are copy-paste from unit tests, right? ;-)

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1121,6 +1121,184 @@ sudo rmmod xocl
+});
+```
+
+for more example see [this test case](../test/on-device/xocc/simple_tests/partion_ndarray.cpp) or [this one](../test/on-device/xocc/edge_detection/edge_detection.cpp)
````
```suggestion
for more examples see [this test case](../test/on-device/xocc/simple_tests/partion_ndarray.cpp) or [this one](../test/on-device/xocc/edge_detection/edge_detection.cpp)
```

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -80,9 +86,9 @@ namespace partition {
       TODO: Deal with multi-dimension array. Now, since we can only deal with
```
Is this TODO still valid?

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -145,8 +150,62 @@ namespace partition {
+  using init = std::initializer_list<typename sub_type::init>;
+
+  template<typename InitTy, typename Lambda>
+  static inline constexpr bool on_each(type& self, InitTy&& other, Lambda&& L) {
```
Curious `on_each` name. It looks like a `any_of` on a `std::ranges::zip_transform_view`.
Add a top-level comment.

Ralender replied:
there is a built-in fast exit path but it is often unused.

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
   /// Construct an array
-  partition_array() {
+  partition_ndarray() {
     // Add the intrinsic according expressing to the target compiler the
```
While you are here try to fix this sentence

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
+  static constexpr bool is_layout_compatible =
+      std::is_same_v<elem_type,
+                     typename std::remove_reference_t<OtherTy>::elem_type>
+          &&std::is_same_v<dim_type,
```
Strange indent/spacing

Ralender replied:
thats what clang-format choose

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
-    /// TODO: Find a way to specialize this with a safer
-    /// implementation when the size of src is at least constexpr
-    std::copy_n(std::begin(src), Size, begin());
+  /// Determine if on other type has the same underlying type and dimension and
```
```suggestion
  /// Determine if another type has the same underlying type and dimension and
```

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
+
+  /// Construct from an N-dimension std::initializer_list
+  partition_ndarray(typename recursive_type::init i) : partition_ndarray() {
+    recursive_type::on_each(elems, i,
```
Now I understand the role of this function. :-)
Perhaps `recursively` should appear in the name?
`for_each_recursively`?

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
+                            [](auto &self, auto other) { self = other; });
   }
 
+  /// Construct from an other partition_ndarray with the same type and
```
```suggestion
  /// Construct from another partition_ndarray with the same type and
```

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
+  /// Construct from an other partition_ndarray with the same type and
+  /// dimensions
+  template <typename OtherTy,
+            typename std::enable_if_t<is_layout_compatible<OtherTy>, int> = 0>
```
All these are just C++20 concepts

Ralender replied:
this really isn't a reusable concept so I think SFINEA is more convenient. but I guess you disagree ?

keryell replied:
I was thinking more to 
```c++
template <typename OtherTy>
  requires is_layout_compatible<OtherTy>
```
not giving a name to such a concept.
Or actually you could just make `is_layout_compatible` the named concept instead of a variable. :-)

keryell commented on sycl/include/sycl/ext/xilinx/fpga/partition_array.hpp:
```diff
@@ -165,103 +224,128 @@ namespace partition {
 
-
-  constexpr const ValueType& operator[](std::size_t i) const {
+  constexpr decltype(auto) operator[](std::size_t i) const {
```
Since you are using `decltype`auto, is it not possible to merge these 2 member functions, the const and non-const one?

Ralender replied:
sadly no. we would need the deducing this proposal. for that

keryell commented on sycl/test/on-device/xocc/simple_tests/partion_ndarray.cpp:
```diff
@@ -0,0 +1,123 @@
+// REQUIRES: xocc
```
At some point we need to replace any `xocc` by `vitis` while adapting the uppercase according to the context.

Ralender replied:
agreed but not in this patch

keryell commented on sycl/test/on-device/xocc/simple_tests/partion_ndarray.cpp:
```diff
@@ -0,0 +1,123 @@
+  // Launch a kernel to do the summation
+  q.submit([&](sycl::handler &cgh) {
+    // Get access to the data
+    auto a_a = a.get_access<sycl::access::mode::write>(cgh);
```
Replace all the accessors with a SYCL 2020 style?

keryell commented on sycl/test/on-device/xocc/simple_tests/partion_ndarray.cpp:
```diff
@@ -0,0 +1,123 @@
+    auto a_c = c.get_access<sycl::access::mode::read>(cgh);
+
+    // A typical FPGA-style pipelined kernel
+    cgh.single_task<class add>([=]() {
```
```suggestion
    cgh.single_task<class add>([=] {
```

keryell commented on sycl/test/on-device/xocc/simple_tests/partion_ndarray.cpp:
```diff
@@ -0,0 +1,123 @@
+    cgh.single_task<class add>([=]() {
+      // Use an intermediate automatic array
+      sycl::ext::xilinx::partition_ndarray<
+          int, sycl::dim<2, 3, 2>, sycl::ext::xilinx::partition::complete<>>
```
`sycl::dim` is an extension and thus should be defined in another namespace?

Ralender replied:
thecnically yes but no one wants to write
```
sycl::ext::xilinx::partition_ndarray<
          int, sycl::ext::xilinx::dim<2, 3, 2>, sycl::ext::xilinx::partition::complete<>>
```

the verbosity is already out of control for extentions.....

keryell replied:
Or `sycl::dim` could be just a `constexpr` normal `sycl::range` instead of having a new type?

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -21,19 +21,114 @@
 import shutil
 import subprocess
+import sys
+from tabnanny import check
```
Warf, a TAB nanny! :-)
But where is it used?

Ralender replied:
its not

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -21,19 +21,114 @@
 import tempfile
 
+# This pipeline should be able to do any promotion O3 is capable of
+# and some more control flow optimizations then strictly necessary.
```
```suggestion
# This pipeline should be able to do any promotion -O3 is capable of
# and some more control-flow optimizations than strictly necessary.
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -238,81 +345,135 @@ def _downgrade(self):
-            "-o", self.downgraded_ir
+            "-o", downgraded_ir
         ]
-        self._dump_cmd("04-downgrade.cmd", args)
```
So now we do not longer have the phase numbering in command names?

Ralender replied:
phase numbers are still there but they are now automatic

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -238,81 +345,135 @@ def _downgrade(self):
+
+
+class VXXCompilationDriver(VitisCompilationDriver):
+    def __init__(self, arguments):
```
Add top level `"""description"""`

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -351,78 +512,126 @@ def _link_kernels(self):
+        return xclbin
 
+class IPExportCompilationDriver(VitisCompilationDriver):
+    def __init__(self, arguments):
```
Top level description

keryell reviewed (changes requested):
Quite interesting!

Commit:
[SYCL] Enable compilling the sycl runtime library in C++17 down from C++20

Commit:
[SYCL][VXX] Rename Xilinx FPGA extentions to snake_case

Apply suggestion

keryell reviewed (commented):
Quite cleaner.

Commit:
[VitisIP] Enable using the C++ system standard library when compilling for vitis-ip.

Note: many components of the standard library will not work.

keryell reviewed (commented):
Interesting.
You could rebase this commit to fix the spelling of the commit message and add some comments in the code because the changes look out-of-the-blue if we do not know we need somehow to make the C++ (x86?) standard library available.
[VitisIp] → [VitisIP]
standare → standard

Commit:
[VXX] Address review comments

Ralender commented:
I rebased as requested

keryell reviewed (approved):
Thanks.
